format-common/src/main/java/io/cdap/plugin/format/output/DelegatingRecordWriter.java

```diff
@@ -34,14 +35,28 @@
 public class DelegatingRecordWriter<K, V> extends RecordWriter<NullWritable, StructuredRecord> {
   private final RecordWriter<K, V> delegate;
   private final Function<StructuredRecord, KeyValue<K, V>> conversion;
+  private final Function<StructuredRecord, KeyValue<K, V>> header;
+  private boolean isHeader;
 
   public DelegatingRecordWriter(RecordWriter<K, V> delegate, Function<StructuredRecord, KeyValue<K, V>> conversion) {
+    this(delegate, conversion, null);
+  }
+
+  public DelegatingRecordWriter(RecordWriter<K, V> delegate, Function<StructuredRecord, KeyValue<K, V>> conversion,
+                                @Nullable Function<StructuredRecord, KeyValue<K, V>> header) {
     this.delegate = delegate;
     this.conversion = conversion;
+    this.header = header;
+    this.isHeader = true;
   }
 
   @Override
   public void write(NullWritable key, StructuredRecord value) throws IOException, InterruptedException {
+    if (isHeader && header != null) {
+      KeyValue<K, V> headerVal = header.apply(value);
+      delegate.write(headerVal.getKey(), headerVal.getValue());
+      isHeader = false;
```

Comment:
Should we return after this line since L61 is writing this record again?

Reply:
Also, will this work if there are multiple splits?

Reply:
can't return because it needs to write both the header and the data row.

There is one RecordWriter per output file -- it isn't re-used across partitions so it won't skip a header.

Reply:
Just confirmed offline - it writes the header based on the record schema to the file, then it writes the actual data in the record. So no need to return.

Better to put a comment to avoid confusion in the future